design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+* Arbitrary WHERE clauses should be allowed.
+* Examples:
+  * `SELECT … FROM table WHERE x = ‘x’ AND y = ‘y’`
+  * `SELECT … FROM table WHERE str LIKE ‘%substr%’`
+  * `SELECT … FROM table WHERE x < y`
+
+### Tier 3 - Aggregates
+
+Aggregations are central to extracting value from datasets.
+
+While materialized views in KSQL may already be defined using aggregates, it will still be very useful to perform further aggregations over them via interactive queries. For example, users may want to maintain a materialized view summarized at the minute level but then summarize it further down to the hour level to power a frontend visualization that dynamically supplies a date range:
+
+* All aggregates supported by continuous queries should be supported by interactive queries.
+* Examples:
+  * `SELECT count(*) FROM table WHERE ROWKEY < 10`
+  * `SELECT day(hour), sum(count) FROM table GROUP BY day`
+  * `SELECT sum(count) FROM table WHERE ts >= ‘2019-01-01’ AND ts < ‘2019-01-02’`
+
+### Tier 4 - Joins
+
+Joins are also fundamental to deeply analyzing datasets and the relationships between various entities. They will also likely be the most complex to support in a clustered environment, which is why they’ve been put in a separate and final tier. As with aggregates, users may already use joins in their materialized view definitions, which may partially mitigate the cost of their absence within the context of interactive queries.
+
+For these reasons, we should probably not focus much on joins until the previous tiers have been delivered.
+
+* Without relatively robust indexing, it will probably only be initially feasible to join on row keys.
+* Examples:
+  * `SELECT … FROM t0 JOIN t1 ON t0.ROWKEY = t1.ROWKEY …`
+  * `SELECT stats_1m.count AS last_minute, stats_60m.count AS last_hour FROM stats_1m JOIN stats_60m ON stats_1m.ROWKEY = stats_1h.ROWKEY`
+
+## User experience
+
+The user experience for interactive queries should be quite simple: if a user issues a query against a KSQL table, that query should return a finite set of rows as quickly as possible.
+
+Additionally:
+
+* All interactive queries must be issued via the REST API.
+* Any interactive query can target any KSQL server.
+* Key movement should be handled transparently: if a query launches and a targeted key physically moves, KSQL should redirect the query to the appropriate KSQL server to obtain the correct result.
+* For interactive queries against windowed queries, the user may specify a time range via the `WINDOWSTART` and `WINDOWEND` psuedocolumns. For example, `SELECT * FROM windowed_aggregate WHERE ROWKEY = 'rowkey' AND WINDOWSTART >= start_ts AND WINDOWEND < end_ts`
+
+## Open questions
+
+1. How will KSQL differentiate between interactive and streaming queries? Should the same differentiation mechanism be applied to both transient queries and CTAS/CSAS statements?
```

Comment:
Are we planning to discuss this here or in a different KLIP? It'd be nice to summarize the different options brainstormed offline for easier evaluation/contrasting.

Reply:
@vcrfxia great idea, I will enumerate all proposed options here on the KLIP.